finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/SamplingTracer.scala

```diff
@@ -0,0 +1,159 @@
+import com.twitter.util.events.Sink
+import com.twitter.util.{Time, Throw, Try}
+import scala.util.control.NoStackTrace
+
+private object Json {
+  import com.fasterxml.jackson.annotation.{JsonTypeInfo, JsonInclude}
+  import com.fasterxml.jackson.core.`type`.TypeReference
+  import com.fasterxml.jackson.databind.{ObjectMapper, JavaType, JsonNode}
+  import com.fasterxml.jackson.databind.annotation.JsonDeserialize
+  import com.fasterxml.jackson.module.scala.DefaultScalaModule
+  import java.lang.reflect.{Type, ParameterizedType}
+
+  // Note: This type is a just a convenience for deserialization in other
+  // other Event.Type constructions, but we actually require it for Trace
+  // because we're using Jackson's default typing mechanism for Annotation.
+  // If we use a Map, somewhere in Jackson's type resolution the type of
+  // Annotation is forgotten, and it is passed into the type resolver as an
+  // Object. Defining this Envelope preserves the type information.
+  @JsonInclude(JsonInclude.Include.NON_NULL)
+  case class Envelope(
+                       id: String,
+                       when: Long,
+                       // We require an annotation here, because for small numbers, this gets
+                       // deserialized with a runtime type of int.
+                       // See: https://github.com/FasterXML/jackson-module-scala/issues/106.
+                       @JsonDeserialize(contentAs=classOf[java.lang.Long]) traceId: Option[Long],
+                       @JsonDeserialize(contentAs=classOf[java.lang.Long]) spanId: Option[Long],
+                       data: Annotation)
+
+  val mapper = new ObjectMapper()
+  mapper.registerModule(DefaultScalaModule)
+
+  // Configures the mapper to include class information for Annotation.
+  object TypeResolverBuilder
+    extends ObjectMapper.DefaultTypeResolverBuilder(ObjectMapper.DefaultTyping.NON_FINAL) {
+    override def useForType(typ: JavaType) =
+    // Note: getRawClass would be an Object if not for `Envelope`.
+      typ.getRawClass == classOf[Annotation]
+  }
+
+  mapper.setDefaultTyping(
+    TypeResolverBuilder
+      .init(JsonTypeInfo.Id.CLASS, null)
+      .inclusion(JsonTypeInfo.As.WRAPPER_ARRAY))
+
+  def serialize(o: AnyRef): String = mapper.writeValueAsString(o)
+
+  def deserialize[T: Manifest](value: String): T =
+    mapper.readValue(value, typeReference[T])
+
+  def deserialize[T: Manifest](node: JsonNode): T =
+    mapper.readValue(node.traverse, typeReference[T])
+
+  private[this] def typeReference[T: Manifest] = new TypeReference[T] {
+    override def getType = typeFromManifest(manifest[T])
+  }
+
+  private[this] def typeFromManifest(m: Manifest[_]): Type =
+    if (m.typeArguments.isEmpty) m.runtimeClass else new ParameterizedType {
+      def getRawType = m.runtimeClass
+      def getActualTypeArguments = m.typeArguments.map(typeFromManifest).toArray
+      def getOwnerType = null
+    }
+}
+
+object SamplingTracer {
```

Comment:
I replaced `c.t.f.zipkin.thrift.SamplingTracer` with a `SamplingTracer` that extends `c.t.f.zipkin.core.SamplingTracer` with a default constructor. This is because the [service loader](https://github.com/twitter/finagle/blob/develop/finagle-zipkin/src/main/resources/META-INF/services/com.twitter.finagle.tracing.Tracer) loads `c.t.f.zipkin.thrift.SamplingTracer`.

Now I realize I can replace `c.t.f.zipkin.thrift.SamplingTracer` with `c.t.f.zipkin.thrift.ScribeZipkinTracer` and load that with the service loader.
I'm not familiar with the intricacies service loading, do you think that will break anything?

Actions speak louder than words so maybe this commit explains this better than I do: https://github.com/twitter/finagle/pull/513/commits/262614b8e5a6d3d005aeb62aebece1565234ce6f